### ISSUE TITLE
CS2 Discussion: Features: let/const/var

### ISSUE THREAD

**Issue by coffeescriptbot** [jashkenas/coffeescript#4900]
_From @DomVinyard on July 11, 2016 12:13_

Personally, I believe these should be untouched. I can see that there is a case for implementing _const_ in almost any language remit. But not coffeescript.

_Copied from original issue: coffeescript6/discuss#1_

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 12, 2016 0:24_

Whether we have `let` and `var` or not is a different question than whether we have `const`. The first two are related to scope, and `const` is about reassignment.

We should keep this discussion about `let` and `var`, and move `const` to a new issue.

Should we even keep CoffeeScript's nonlocal assignment rule? If were making breaking changes, we should consider changing that as it was quite controversial - all assignments are currently nonlocal. The name lookup starts locally, and works outwards until it finds the name and reassigns it. If it doesn't find the name, the assignment is local.

In Python, all assignments are just local, which is what you normally want to happen. Lexical scope exists to keep variables locally defined. You never have to worry about clobbering something in the global scope with a local variable like you do in JS and CoffeeScript.

The problem with Python is that it uses `nonlocal` statements when you want to reassign names in outer scopes, the way CoffeeScript does by default, and we really want to avoid `var` and `let` and any other assignment keywords. They're especially awkward in a language where everything is an expression.

## Bikeshedding

I think all assignments should be local, and we should have a variable available inside each function, just like `arguments`, called `outer`, that's a reference to the function's outer scope. 

You would use `outer` the same way we use `window` in the browser to do global assignments:

```
outer.x = 1
```

If you had functions nested inside of each other, each lexical scope would have its own `outer`, so it would naturally be recursive. You could do `outer.outer.outer...` all the way to the global scope.

Assignments should have always been local, and it's natural to extend the idea behind `window` to all scopes. Then you don't need assignment keywords, can access any scope, and it works naturally as an expression.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 12, 2016 4:55_

`const` is now often the default in ES6 code. Making it convenient to use would be a plus. But not necessarily a priority. 

My hunch is the `outer` proposal, while interesting, would be too far a departure to take on right away.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 12, 2016 4:58_

I don't think there's much reason to use `let` since coffeescript's `var`'s have (most of) the same benefits; they are block-scoped. 

If adding a `const` keyword isn't too hard, it might be worthwhile. 

An alternative would be to have all variables by `const` by default, and allow explicit `var` or `let` declarations. I assume this would be more work, as the scoping code coffeescript undertakes would be mostly-removed.

**Comment by coffeescriptbot** [comment]
_From @DomVinyard on July 12, 2016 8:24_

> You could do outer.outer.outer... all the way to the global scope.

That sounds like an absolute nightmare. Hard-coding your depth in the scope.

It will be very easy to reimagine all of coffeescript's features under the mandate of 'allowing breaking changes' - but I think there is a good case for picking a few bits of low hanging fruit and then addressing the rest once we're up and running. let/var/const would be a good candidate for leaving alone, i think.

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 15, 2016 16:11_

I don't see a problem with absolute scope references, when scope is already lexical. Still, we have more worthwhile stuff to discuss, so yeah, forget it.

**Comment by coffeescriptbot** [comment]
_From @JimPanic on July 19, 2016 9:17_

I really want to see `const` in this new endeavour in one way or another. Maybe not in the first iteration, but definitely when targeting ES6. This is a key feature for me (and I guess a lot of other people as well) in ES6.

Regarding scoping, I'm in favour of leaving lexical scope as-is as well. In the future it might be a good idea to make `let` the default, `const` easy to use and `var` the least common thing syntactically.

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 19, 2016 13:20_

Hi @JimPanic, welcome.

> Regarding scoping, I'm in favour of leaving lexical scope as-is as well.

You misunderstood the problem I was trying to address. CoffeeScript will always have lexical scope as-is. No one wants to change that. It's the way CoffeeScript mandates that all assignments are nonlocal that's the issue. It's always been a pretty controversial feature, and one of the most popular criticisms of CoffeeScript. For an example of the feature, here's two seemingly similar blocks of code, the first is CoffeeScript, the second is Python. They do different things though:

``` coffee
    x = false

    do ->
        x = true
        console.log x # true

    console.log x     # true 
```

In CoffeeScript, the outer `x` is clobbered by the inner assignment. Whereas, without an explicit `nonlocal x` statement, Python will never clobber names outside the local namespace.

``` python
    x = False

    def f():
        x = True
        print x       # True
    f()

    print x           # False
```

We can avoid clobbering `x` in CoffeeScript too, if we introduce the name into the local namespace as a parameter, but you still have to know that that name exists in some outer scope to preempt the clobbering.

``` coffee
    x = false

    do (x) ->
        x = true
        console.log x # true

    console.log x     # false 
```

Lacking any real evidence, I can just say from experience with Python that nonlocal assignment is pretty rare. You almost always want assignments to be local. I assume @jashkenas only went with nonlocal assignment to avoid having to have declarators. There's no other sane reason to do it.

If we're going to change the language design to introduce declarators, then we should definitely make assignment local by default and have a declarator for nonlocal assignment. We don't need Python's `global` declarator as we have `window.x` already.

If `let` makes a variable block-scoped and `const` makes the name un-reassignable and block-scoped, neither does what we need.

Even if we keep assignments nonlocal, if we introduce declarators, we would really need declarators for `local`, `const` and `local const`.

The benefit of having block-scope in CoffeeScript is slim. We can always create a new scope with `do ->`, and any variables in the function will be effectively block-scoped (again, assuming we don't clobber anything). Having `let` would allow us to do this:

``` coffee
if foo
    let x = true
    let y = false
```

But, we can already do this anyway:

``` coffee
if foo then do ->
    x = true
    y = false
```

JavaScript became considerably more complicated by introducing `let`, and it was only introduced to fix different problems to the ones CoffeeScript has. We don't have much use for it.

I personally think `outer.x`is better than `nonlocal x` _because_ it's explicit about which scope it references, and follows logically from `window.x`. Still, I'm happy to hear other ideas.

Taking declarators from JavaScript requires changing them to fit CoffeeScript's nonlocal assignment feature, which may in turn require any declarators to be renamed for their new semantics, so we'd basically need to define our own declarators from scratch. And we really should try and do everything we can to avoid introducing declarators, given that they have been rejected from CoffeeScript for seven years, and they make everything look like Java.

``` coffee
jQuery ->

    local const square = (x) -> x * x
```

**Comment by coffeescriptbot** [comment]
_From @JimPanic on July 19, 2016 13:58_

@carlsmith You are right, I did misunderstand.

I always thought CS was favouring the innermost scope in assignments. Lucky me it didn't bite me so far.

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 19, 2016 14:3_

@JimPanic :)  It's not obvious what CoffeeScript is doing, especially when it seems to be doing what you'd expect, coming from other languages. It doesn't bite that often, but when it does, it can be totally baffling.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 20, 2016 20:25_

Thanks very much @carlsmith for the terrific explanation there. Didn't understand that bit myself so well till now 😄 

I do agree that it would likely be too ambitious to change how scoping works, unless we can simplify the coffeescript compiler by staying closer to JS's behavior.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 20, 2016 20:27_

> [const] is a key feature for me

@JimPanic can you elaborate? I intuitively feel the same way – I enforce `const`-by-default in my eslint – but curious to hear if there are additional reasons why `const` is useful.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 20, 2016 20:39_

Any thoughts on `const` being used like so:

``` coffee
x := 7  # const x = 7
```

And all other `=` being treated as they are today, either using `var` or `let`?

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 20, 2016 20:43_

> Personally, I believe these should be untouched. I can see that there is a case for implementing const in almost any language remit. But not coffeescript.

@DomVinyard re-reading this, I'm not sure I understand what you originally meant. Can you clarify?

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 20, 2016 21:29_

Adding a constant operator is easier than adding declarators. It changes the language less. The operator is ok, but will get a bit messy when its `x :=? 7`. If it was a short word, but still a normal operator, it wouldn't look as cryptic: `x be 7` and `x be? 7` looks more like regular CoffeeScript. I don't think it should be called `be` though. Just, if there was a good, short word, that would be better than having a `:=?` operator  :)

**Comment by coffeescriptbot** [comment]
_From @JimPanic on July 22, 2016 8:38_

@rattrayalex It helps avoiding so many bugs and enables a more declarative way of coding, which in turn helps keep code readable and (more) testable. How do you enforce `const` by default? Does your linter settings just yell at you when you do not declare something as const? :D

**Comment by coffeescriptbot** [comment]
_From @JimPanic on July 22, 2016 8:39_

@carlsmith I was thinking the same; a different assignment operator would go well with the CS syntax as it is now, since there are no declarator keywords. Why the question mark in `:=?` ?

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 22, 2016 12:55_

@JimPanic - My bad: I got it back-to-front (I rarely use the feature this way), but you have an _existential assignment_ operator in CoffeeScript that looks like `?=`, so if we have a _constant assignment_ operator like `:=`, we'd need an _existential constant assignment_ operator, which would end up looking like `?:=`.

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 22, 2016 13:26_

We could figure out from static analysis whether a name is assigned more than once, and make names that are never reassigned constants automatically, but expect JavaScript engines already do this, so any efficiency gains would likely be negligible. More importantly, people want explicit constant assignment/declarations to communicate that constraint to other programmers. Given that it's a popular CoffeeScript feature request, and already part of ES6, explicit constants should really be supported somehow.

There's an argument for getting rid of the `is` and `isnt` operators from CoffeeScript. CoffeeScript used to have an `aint` operator that compiled to `!==`, but it got dropped in favour of just having the current `isnt` alias, but that's confusing. People often shorten `a is not b` to `a isnt b`, expecting them to mean the same thing, like they do in English, but `a is not b` compiles to `a == !b`, while `a isnt b` compiles to `a !== b`.

If we just kept the current `==` and `!=` CoffeeScript operators (which compile to `===` and `!==`), and got rid of `isnt` altogether, then `is` would be a good constant assignment operator.

``` coffee
    i = 0
    pi is 3.141
```

Then `a ?is b` would work as an existential assignment operator.

If we do a backwards incompatible CoffeeScript, we should definitely fix anything that is accepted as broken and easy to fix. Removing `is` and `isnt` would be trivial. We already have conventional, unproblematic `==` and `!=` operators. And `is` would be a _really_ nice constant assignment operator.

It's also worth mentioning that Python has an `is` operator. It's used to test whether two expressions evaluate to the exact same object in memory.

``` python
    a = b = []
    a is b         # True

    a, b = [], []
    a is b         # False
```

Using `is` for assigning constants would be a bit different, but still similar to Python - they both kind of mean `is actually` rather than `equals`.

**Comment by coffeescriptbot** [comment]
_From @DomVinyard on July 22, 2016 14:25_

@rattrayalex 

Coffeescript succeeded, at least partially, because it traded out efficiency optimisations for simplicity in a bunch of contentious but elegant ways. Not distinguishing between _let_ and _const_ fits that core simplicity brief.

Perhaps this means that a few edge-projects will have to use vanillaJS for certain classes of application because the overhead of trading out lets for consts would trip the project from acceptable into unacceptable but I think that is a reasonable cost to keep things simple.

Avoiding having to add a new operator is worth fighting hard for in every case.. if CS is to retain its CSiness.

**Comment by coffeescriptbot** [comment]
_From @DomVinyard on July 22, 2016 14:40_

@carlsmith Firm -1 from me for redefining _is_.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 22, 2016 16:24_

> How do you enforce const by default? Does your linter settings just yell at you when you do not declare something as const?

Yes – it's an ESLint setting to always use const unless you redefine the variable later. 

> we'd need an existential constant assignment operator, which would end up looking like ?:=.

I don't think an existential constant assignment operator makes sense 😉. `?=` is used for reassignment, which isn't possible with `const`.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 22, 2016 16:55_

> explicit constants should really be supported somehow.

I'm increasingly convinced of this as well personally. Should probably add to #8 . 

> Avoiding having to add a new operator is worth fighting hard for in every case.. if CS is to retain its CSiness.

I disagree with this sentiment, actually. While maintaining a decent degree of backwards-compatibility would be nice, I don't think we should shy away from adding new features, especially those that are part of ES6. The number one purpose of this org is to ensure that CoffeeScript isn't _behind_ ES6 at all – that anything\* you can do conveniently in JavaScript, you can do at least as conveniently in CoffeeScript. 

Furthermore, language innovation is a core aspect of the "CSiness" of CoffeeScript IMO. Jeremy combined the most convenient language features he could find from other languages, made it easy to write good javascript, and added some new ideas into the mix as well. 

We should certainly think really hard about whether any new features are worth doing – and focus on compatibility with ES6 first – but I also don't think it's something we should shy away from.

\* Anything that you should be doing, anyway. Making "the bad parts" and generally poor programming practices less convenient, or even impossible, is a good thing.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 22, 2016 16:59_

> Firm -1 from me for redefining is.

I personally agree. However, to keep a healthy environment where people feel comfortable voicing new ideas they might not be 100% comfortable with yet, let's try to back up -1's with a few calm reasons. 

In this case, my reason is primarily difficult of porting – redefining a language keyword to something completely different could make updating quite difficult, as you'd have to both translate all existing `=`s to `is`, and only once that is 100%-for-sure-done across your whole project, change most of your `=`s to `is`. 

I also happen to think it'd be a bit less readable than something with symbols, given the ubiquity of the `=` for assignment across programming languages.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 22, 2016 17:1_

My understanding is the primary objection to `:=` is the conditional assignment scenario, which as I outlined above probably doesn't apply. Are there others?

I think the leading alternative at this point would probably be to just pass `const` through, or come up with a similar (ideally shorter) word. 

Thoughts?

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 22, 2016 22:0_

> I don't think an existential constant assignment operator makes sense 😉. ?= is used for reassignment, which isn't possible with const.

I didn't think about that. Sorry about the noise then. I mentioned not using that feature much when I misspelled the operator `:|`

I'm with @rattrayalex on adding new features, but feel pretty strongly that we should stick to things that are just sugar for well established idioms or that have already been standardised by ECMA. CoffeeScript should make one big breaking change, then treat TC39 as our standards process. Now that CoffeeScript has inspired JavaScript to evolve, we can co-evolve with JavaScript in a way Jeremy couldn't in 2009.

So, yeah - sorry to ramble - _assuming_ there will be no need to have an existential constant assignment operator, then I would personally be ok with `:=`, but still wish it was prettier.

```
square := (x) -> x * x
```

I still feel strongly that we should avoid declarators if at all possible. It changes the language, and we use assignment expressions in lots of places where they need to be concise, not just in standalone statements..

Also, do we need a story for constant function parameters, or can you just not define a parameter as a constant?

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 23, 2016 2:11_

> but still wish it was prettier.

Agreed. 

> remember that we use assignment expressions in lots of places where they need to be concise.

Would you be willing to provide a few quick examples for illustrative purposes?

> or can you just not define a parameter as a constant?

Well, you can't in es6, so we'd really be going above and beyond if we implemented it in cs6. It's not an unappealing feature imo but probably not something to worry about yet. 

> CoffeeScript should make one big breaking change

Amen. 

> then treat TC39 as our standards process

I really need to learn more about TC39 😃

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 23, 2016 2:57_

> Would you be willing to provide a few quick examples for illustrative purposes?

I meant using assignment expressions like this:

``` coffee
if (stuff = getStuffOrReturnNull) then use stuff

employees = [
    ali = new Employee "Ali"
    bob = new Employee "Bob"
]

```

Because assignments are expressions, you can write one anywhere you can write an expression. It can be helpful when you want to make assignments conditionally, and lets us use some shorthand expressions for initialising variables.

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 23, 2016 3:10_

On the TC39 comment: I really just meant that we should follow the ECMAScript standards when introducing features that go beyond sugar for common JS idioms.

We have the luxury these days of a standards process that moves quickly, releases often and develop editions in parallel. It has some of the best minds in our community, and all the vendors are on board. You can get involved through ESDiscuss, and influence the spec, but for the most part, we don't need to. It works.

If Jeremy had followed the spec, we wouldn't have the problem we now have with incompatible classes. Obviously, in 2009, following the spec would have made CoffeeScript suck.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 23, 2016 3:13_

Wow, I've never seen the `employees` code before. Looks awesome though. 

https://github.com/michaelficarra/CoffeeScriptRedux/wiki/Intentional-Deviations-From-jashkenas-coffee-script#intentional-deviations might be relevant here once we get into the weeds, especially if we decide to build on CSR. 

Thanks for sharing the examples! Helpful.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 23, 2016 3:14_

Re; TC39, thanks for elaborating - helpful. To clarify, do you propose:
1. Implement all things agreed by TC39
2. Set up a similar organization with similar bylaws, and use that to determine the course of the language. 

(I'm mostly 👍 to both)

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 23, 2016 3:55_

Some thoughts on an alternative route to `:=`. 

There are three things that we typically do with `=`:
1. Declaration (`const myVar`, `let myVar`, `var myVar`)
2. Assignment (`let myVar = 3`)
3. Reassignment (`let myVar = 3; myVar = 'foo'`)

Reassignment occurs the least often, and is arguably an anti-pattern (though often a practical necessity). `const` prevents reassignment, `let` and `var` do not. `const` requires assignment with declaration, `let` and `var` can be declared without assignment. 

There are a couple routes we could go down: 
1. Use `const` by default, and infer when `let`/`var` should be used instead. 
   Leverage coffeescript's existing Scope parser to check if a variable is ever reassigned; if not, it's `const`. If it's reassigned, it's `var` or `let` (we can decide later). 
   **Pros**: This would be backwards-compatible (modulo some probable bugs) and would result in frequent use of `const`. 
   **Cons**: Since there is nothing visible to the coffeescript programmer, the benefits of using const are essentially eliminated. The work of discouraging reassignment would probably have to fall to a linter. 
2. Use `const` by default, and enforce declarators for `let`/`var`. 
   This would make using `let`/`var` less convenient; the example above would probably look like this:
   
   ``` coffee
   let employees, ali, bob
   employees = [
       ali = new Employee "Ali"
       bob = new Employee "Bob"
   ]
   # later, reassign to different values:
   ali = 5
   bob = 'foo'
   employees = 'other nonsense'
   ```
   
   (the `let` would be unneeded if the variables were never reassigned)
   **Pros**: This would discourage use of reassignable variables, while still allowing it without too much hassle. It should be reasonably tractable to write an automated cs->cs6 transpiler to add the `let`'s, but probably a pain. It also allows use of both `let` and `var` without taking a stance. 
   **Cons**: Declarations are still ugly. Might be too prescriptive. 
3. Use `const` by default, and enforce special syntax for reassignment (eg; `^=`, the update-operator). 
   This is pretty outlandish – I don't know of any other language that has a different syntax for assignment vs reassignment – but I think it's pretty interesting. Imagine reading a program and immediately knowing whether `foo = 'bar'` is writing a variable or overwriting a variable. 
   It could look like this:
   
   ``` coffee
   foo = 1
   foo ^= 2
   ```
   
   Or, perhaps:
   
   ```
   foo := 1  # becomes `let foo = 1`
   foo ^= 2 # becomes `foo = 2`
   bar = 3   # becomes `const bar = 3`
   ```
   
   Or this could be done in tandem with option 2:
   
   ```
   let foo = 1  # becomes `let foo = 1`
   foo ^= 2 # becomes `foo = 2`
   bar = 3   # becomes `const bar = 3`
   ```
   
   **Pros**: Would strongly discourage, but allow, reassignment. Would be extremely explicit about what is assignment, what is reassignable-assignment, and what is reassignment. Possibly a genuine language innovation. Can get rid of CoffeeScript understanding of scope entirely.
   **Cons**: Inconvenient, ugly, foreign. 
4. As mentioned above, default to `var` and allow `const` declarations through `:=`. 
   **Pros**: Much more backwards-compatible, still allows `const` for those who really want it.
   **Cons**: Discourages the thing you should really be doing; the right thing and the good thing aren't the same.
5. Default to `const`, and allow `var` or `let` declarations (but only one of the two) through `:=`. 
   **Pros**: Relatively simple, avoids declarators, could possibly get rid of CoffeeScript understanding of scope, makes the right thing the easy thing. 
   **Cons**: Still a little ugly, not backwards-compatible, have to make a choice between `let` and `var`. 

Thoughts? Any other ideas?

**Comment by coffeescriptbot** [comment]
_From @ozjd on July 24, 2016 17:39_

To keep things very CoffeeScript like, I would use the following: (in this order)
`const`: When a variable is named UPPERCASE eg `VAR = ...` (It's very important to note that this seems to be the common suggestion for coffeescript, and is how most people already write an implied constant)
`var`: When an object is redefined (eg, contains a `.` as in `window.var = ...` or `[]` as in `window['var'] = ...`
`let`: All other times.

Examples:
`@foo = 'bar'` would become `let this.foo = 'bar'`
`@FOO = 'bar'` would become `const this.foo = 'bar'`
`obj::foo = 'bar'` would become `let obj.prototype.foo = 'bar'`
`obj::FOO = 'bar'` would become `const obj.prototype.foo = 'bar'`
`obj.prototype.foo = 'bar'` would become `var obj.prototype.foo = bar`
`obj.prototype.FOO = 'bar'` would become `const obj.prototype.FOO = 'bar'`

I'm sure there are other cases that I haven't thought of, but I'd imagine that this would also probably break very few existing scripts. Thoughts?

**Note: I think the most important part is automatically mapping UPPERCASE variable names to a `const`.**

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 24, 2016 18:35_

I didn't quite understand the code above, but just to be clear: Deleting `ali` and `bob` in this code would not change what the array named `employees` evaluated to:

``` coffee
employees = [
    ali = new Employee "Ali"
    bob = new Employee "Bob"
]
```

The code creates three names:
- `employees: [<Employee0>, <Employee1>]`
- `ali: <Employee0>`
- `bob: <Employee1>`

Deleting `ali` would decrement the reference count on `<Employee0>`, but it would not be garbage yet, as it's still referenced by `employees[0]`.

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 24, 2016 18:37_

I already commented about this on the main CoffeeScript repo, but beings as it's being addressed here as well...

`-1` on forcing constants to be uppercase. Most functions will be constants, and we don't want every function to have a `LOUD_NAME`. We also wouldn't be able to distinguish between class names and the names of regular functions if they all have to be uppercase. We also use camelCase a lot in CoffeeScript, and would be forced to use underscores if we only have uppercase constant names.

A constant assignment operator has been discussed elsewhere, so you'd do something like:

``` coffee
i = 0
pi := 3.141
```

A better looking operator would be nice, but it's a much less radical change at least.

**Comment by coffeescriptbot** [comment]
_From @ozjd on July 24, 2016 19:52_

Note that I wasn't suggesting that you **MUST** use uppercase for constants, just that UPPERCASE variables would be considered constants.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 24, 2016 17:31_

@ozjd I might recommend you familiarize yourself with ES6 a bit. 

Variable declaration isn't done as a part of property assignment; `let obj.prop` results in `Uncaught SyntaxError: Unexpected token .`. 

`const` is commonly used outside of traditional "constants". For example, AirBnB's style guide (the de facto community standard) recommends using `const` for as many variable declarations as possible.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 24, 2016 17:40_

@carlsmith sorry for my confusing code sample, I didn't mean to delete the variables, just reassign them. I've edited my comment for (arguable) clarity. Does that help at all?

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 24, 2016 17:56_

Ah, one idea I meant to add to this thread that could make the parser simpler and the language more explicit: 

Outlaw implicit `window`. 

Now that ES6 has modules, the value of the convenience of the implicit window has gone way down. It's almost exclusively an antipattern. In some files where I need to use a global, I put this at the top: 

`const { app } = window;  // fake import`

... and otherwise just use `window.someVariable`. 

CoffeeScript already does not allow implicit writes to the global object, but does allow implicit reads: 

``` coffee
thing() ->
  f()  // would essentially call `window.f()`
```

Should we do this if it makes life easier?

**Comment by coffeescriptbot** [comment]
_From @GeoffreyBooth on July 24, 2016 21:05_

I guess an early question to answer is whether we want to give the developer the power to specify when `let` or `const` is output. In other words, what's the goal here:
1. To build the ability for the compiler to have the smarts to know when to use `let`, when to use `const`, and when to use `var`? And current syntax is unchanged.
2. To allow the developer to force a `const` for a particular variable? And the compiler would automatically use `let` or `var` for all others.
3. To allow the developer to force a `let` for a particular variable? And the compiler would automatically use `const` or `var` for all others.
4. To allow the developer to force _either_ a `const` or a `let` for a particular variable? And the compiler would automatically choose the best option for all others.

I assume we at least want to do the first goal, and that should get its own issue to figure out the methods for how the compiler will decide what to do. If we want to do one of the others, perhaps a new thread can be created where we can hash out the desired functionality and new syntax.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 25, 2016 20:28_

Terrific summary @GeoffreyBooth . Personally I'm interested in 1, 2, or 3. I'd prefer 3 but for backwards-compatibility would accept 2. 

`const` exists, and is used, for a reason. That reason goes away if the compiler infers which one you mean, hence my unease with option 1. What's the point of `const` at that point?

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 25, 2016 20:31_

The point in that case would be that the JS engine gets to do all the optimisation, without us having to tell it what to do manually. We get all the benefits, with none of the labour.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 26, 2016 22:05_

What benefits?

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 26, 2016 5:12_

Perf.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 26, 2016 10:48_

Sorry, I should have been more clear. I don't think performance is the reason people use `const` in JavaScript. 

As of [Oct 2015](http://stackoverflow.com/a/33456813/1048433):

> Using `const` is consistently slower in all tests but it is marginal and too close to call.

And was much slower when it was first released. It might be faster in the long term, of course, but for now it's not – especially if you use Babel, and the `const` is turned into a `var` anyway. As such, it probably wouldn't make sense to enable `const` at all if performance is the argument, until there is indeed a dramatic performance boost from using it. At which time, yes, implicit `const` would probably be fine.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 26, 2016 10:48_

My impression is that `const` is used for preventing the class of bugs related to reassigning variables. Eg; from the AirBnB style guide [note on the topic](https://github.com/airbnb/javascript#references--prefer-const): 

> This ensures that you can't reassign your references, which can lead to bugs and difficult to comprehend code.

From [the ESLint docs](http://eslint.org/docs/rules/prefer-const.html): 

> If a variable is never reassigned, using the `const` declaration is better.
> 
> `const` declaration tells readers, “this variable is never reassigned,” reducing cognitive load and improving maintainability.

These benefits don't apply if `const` declaration is invisible and goes away the moment you reassign a variable.

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 26, 2016 11:03_

Yeah, that's the controversial bit. Jeremy could have added constants to CoffeeScript at any point - the compiler knows which vars are reassigned - but didn't. The perf thing is a definite benefit if it works, but as you pointed out, it doesn't seem to. The benefit of explicit constants has to be weighed against complicating the language.

Personally, I think you're right: It's a feature people want, and it's now standardised, while it wasn't when Jeremy decided against it, and the complication is pretty minor, especially for the kind of people CoffeeScript appeals to, so I'd vote to accept a PR if asked.

If we work on the basis that concensus means that nobody strongly objects (as it's impossible to get everyone to agree), then the question is really simple... 

Does anyone feel strongly that we should not add explicit constants?

**Comment by coffeescriptbot** [comment]
_From @GeoffreyBooth on July 26, 2016 11:39_

Personally I think we should do both: the compiler should automatically choose the best of `var`/`let`/`const` for each variable assignment, _and_ we provide developers a way to force a `const` for a particular assignment. I don’t see a need to provide a way to force `var` or `let`.

I like the idea of echoing the `@foo` as our way of designating a constant, by using some symbol other than `@` to designate a constant. My instinct would be to use `!` but that already means “not.” So . . . `^foo`? `%foo`? Is there something standard from other languages?

**Comment by coffeescriptbot** [comment]
_From @carlmathisen on July 26, 2016 11:47_

Agree with @GeoffreyBooth that we don't need to worry about `var` or `let` when writing CS code. I mean, CS takes care of that scope for you under the hood anyway.

Regarding `const` assignment: In normal JS and other languages, you just reference a constant like a normal variable, don't see why CS needs to do this differently. IMHO, it just adds unnecessary complexity to syntax.

Why not just `const foo = 'bar'`?

**Comment by coffeescriptbot** [comment]
_From @JimPanic on July 26, 2016 11:52_

I feel strongly _in favour_ of adding some form of `const` syntax. It does not have to be a declarator keyword. To keep the concise nature of CS, the aforementioned `:=` for `const` is what I would like to see.

Regarding `let` and `var`: since `const` will then not be the default, `let` should be the default behaviour. `var` could be explicitely declared using the `var` keyword.

@carlmathisen: I think the main thing with this is, there are no declaration keywords in CS and it would change the language quite a bit.

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 26, 2016 15:25_

If we're compiling to ES6, we'll definitely end up using `let` in compiled JS (whether we redo CoffeeScript's nonlocal assignment or make assignments local or whatever happens). CoffeeScript currently wraps stuff in self-invoking lambdas to create the same effect with`var`, so we'd definitely use `let` instead these days.

Everyone who's contributed so far seems to agree we should have explicit constants, that compile to `const` and so have ES6 semantics (the name can't be reassigned, but the value can still be mutable). We keep going back to declarators though. I'm pretty strongly against them. They are too bulky:

``` coffee
const factory = (mutator) -> (args...) ->
    mutator (const self = Object.create null), args...
    return self

Employee = factory (self, name, salary) ->
    self.salary = salary
    const self.name = name
    const self.raise = (amount) -> self.salary += amount

const ali = Employee "Ali", 50000
ali.raise 10000
```

CoffeeScript hasn't got declarators, so adding them is a big change. From that example (which I only went with because I use it all the time), `const` is used five times. It makes much more sense to use an operator. Adding an operator doesn't change things as radically, though it does change things substantially still.

``` coffee
factory =: (mutator) -> (args...) ->
    mutator (self =: Object.create null), args...
    return self

Employee = factory (self, name, salary) ->
    self.name =: name
    self.salary = salary
    self.raise =: (amount) -> self.salary += amount

ali =: Employee "Ali", 50000
ali.raise 10000
```

I put the equals first because GitHub kept trying to insert people's names when it's written the other way around.

There's the option of appending something to the beginning or end of a constant, but that seems pretty bulky in practice too.

``` coffee
@factory = (mutator) -> (args...) ->
    mutator (@self = Object.create null), args...
    return self

@Employee = factory (self, name, salary) ->
    self.@name = name
    self.salary = salary
    self.@raise = (amount) -> self.salary += amount

@ali = Employee "Ali", 50000
ali.raise 10000
```

If it looks like a different programming language, it's gone too far, so only an operator seems acceptable. If we could get it down to one character, then it would be better still, but looking at the keyboard, there's nothing left, so something like `=:` seems like the best compromise, but it is a bit ugly still.

We could use unicode, but even in a language for the Web, that's still a big change. Looks good though.

``` coffee
i = 0
pi ▹ 3.141
```

It would be nice to support unicode for people who want to use it, in a subproject ([APL](https://en.wikipedia.org/wiki/APL_%28programming_language%29) Mode). Using stuff like `≠`, `≤`, and `≥` would make code less cryptic and prettier, and isn't hard to enter once you start using <kbd>Alt</kbd> and <kbd>Alt</kbd>+<kbd>Shift</kbd> for a little while.

Given that so many CoffeeScript users like to do language design, we should give plenty of thought to letting people do stuff like configure the compiler, expand macros and walk the AST mutating nodes. Obviously, we can't do everything, but letting people have the dialect they want would be cool.

Sorry for such a long post.

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 26, 2016 15:34_

Just really wanted to add something: There's a long list of abandoned CoffeeScript dialects. If we could find a way to allow people in future to implement a dialect through the official API hooks and compiler configurability, then we could bring all that stuff into one project with a bunch of third-party language extensions and entire domain specific dialects available.

I've been playing with Pratt parsers quite a bit, and they are relatively easy to provide as a library the user can easily extend. It's just a bit awkward without module support, because there's some intrinsically shared state, but I reckon it could be done that way really well. The issue with CoffeeScript for any parser is all the implicit stuff.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 26, 2016 17:19_

@carlsmith your example should read: 

``` coffee
factory := (mutator) -> (args...) ->
    mutator (self := Object.create null), args...
    return self

Employee := factory (self, name, salary) ->
    self.name = name
    self.salary = salary
    self.raise = (amount) -> self.salary += amount

ali := Employee "Ali", 50000
ali.raise 10000
```

recall that properties (eg; `self.name`) are not declared.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 26, 2016 17:24_

Alright, it's definitely sounding like `:=` as `const` is the way we'll go for now. 

If we can use `let` instead of `var` too, cool, but I don't think there's a reason to prioritize that. 

Backwards-compatible, simple, concise. Lets people use it who want to, nobody has to. Should simple to implement in relative terms. 

Sound good?

**Comment by coffeescriptbot** [comment]
_From @dadleyy on July 26, 2016 18:30_

In my ideal version of what we're trying to build here, I'd sooner have the compiler use `ALL_CAPS` as an indicator of const-ness than adding an operator.

It seems that if the compiler is going to be intuitive about two out of these keywords, using an operator like `:=` makes the third an oddball even though they're all related. 

When I write things in all caps (which I do), my intention is to indicate to other people working in my code that that symbol's value is not going to change and I think thats consistent across other languages too (whether or not they have the concept of const-ness).

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 26, 2016 21:03_

@dadleyy have you programmed in ES6 much? `const` is used differently than in most other languages.

**Comment by coffeescriptbot** [comment]
_From @JimPanic on July 26, 2016 23:07_

@dadleyy `const` in JS is _not_ about immutability, which would be indicated in other languages by using all-caps variable names. It is about the fact that a variable declared with `const` can only be assigned to once.

This essay lays out the characteristics of `let`, `var` and `const` in a very comprehensible way: http://raganwald.com/2015/05/30/de-stijl.html

**Comment by coffeescriptbot** [comment]
_From @dadleyy on July 27, 2016 6:10_

I understand it does not prevent users from mutating the value stored, e.g:

``` javascript
const user = {name: "johnny"};
user.name = "bill";
```

But I just think that if someone modifying the contents of a something identified as `const` is bad practice - whether or not it's allowed. Just personal opinion; the majority of my use for this in es6 code is to define string, number, and regex literals:

``` javascript
const MAX_ATTEMPTS = 3;
const INVALID_BUDGET_ERROR = i18n("transactions.invalid_budget");
const VALID_INPUT_RGX = /[A-Za-z]/
```

Wherein the value is almost immutable anyways. Is there an example of a situation where the code uses `const` with an object/array that it does modify that is beneficial to the code? Where are you guys using `const` today and for what?

**Comment by coffeescriptbot** [comment]
_From @JimPanic on July 27, 2016 6:16_

I'm using `const` actually everywhere and make `var` or `let` the exception for when I really need to reassign a variable. I wouldn't want most of my function scope to be cluttered with all-caps variable names. It would look a bit odd, I think. 🤔

**Comment by coffeescriptbot** [comment]
_From @dadleyy on July 27, 2016 6:19_

ah. I feel like having it cluttered by `:=` would be worrysome to though, no? Is the best case scenario here a compiler that detects re-assignment and appropriately decides `let`/`const`/`var` without explicit user involvement?

**Comment by coffeescriptbot** [comment]
_From @JimPanic on July 27, 2016 6:32_

Since the assignment with `:=` only happens once, but usage of variables is probably more often, I wouldn't think so.

My problem with inference of `let`/`const`/`var` is that it completely goes against the advantage of `const` as an explicit marker. You'll never know whether a variable is declared as `const` unless you read the complete code of the current scope.

I do get your point with all-caps const variables and I've used that in Ruby a lot. But it kind of bugged me to write and read all-caps variable names as well.

What do other people think about this?

**Comment by coffeescriptbot** [comment]
_From @JimPanic on July 27, 2016 6:35_

It all boils down to taste, and I'm frank about it: I don't like all-caps words and I find `:=` vs. `=` to look better and coincidently they correlate quite well with the mathematical use of those operators: https://en.wikipedia.org/wiki/List_of_mathematical_symbols#Symbols_based_on_equality .

So basically your proposal would work equally well but I personally don't like it as much. :)

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 27, 2016 6:44_

I'm totally against `SQUARE = (x) -> x * x`. It's a nightmare to have to hold Shift every time you type a name as well.

It also destroys all of our naming conventions, which span many languages.

**Comment by coffeescriptbot** [comment]
_From @dadleyy on July 27, 2016 6:52_

Yeah I didn't realize people were using `const` so much; having too many uppercased symbols would definitely get annoying (though one or two is definitely fine @carlsmith, I wouldn't consider that a nightmare). I'm just worried about departing from coffeescript too much by adding `:=`. If that's what people want and it makes it in, I'm still going to use this new language so my case is pretty moot.

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 27, 2016 7:03_

One or two is fine, but most names are constants.

@rattrayalex - In the example I posted, `Employee` should still be a constant.

Why can we not make properties constant? We can make properties of `window` constant. Every name is really just the property of some scope. I don't like the idea of adopting broken operators. If ES6 has a broken constant operator, then maybe we need to fix it.

**Comment by coffeescriptbot** [comment]
_From @carlmathisen on July 27, 2016 8:46_

> @carlsmith: We can make properties of `window` constant.

Didn't know that - how?

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 27, 2016 9:21_

I just meant that global assignments are to properties of the window object, so outside lexical or modular scope, `const x = 1` and  `const window.x = 1` are equivalent, but the second is apparently illegal.

I've tried to read up on this, as it's a bit confusing - you assume it's as simple as it seems - but the usual docs don't dive into any detail. Gonna search ESDiscuss later, and see what's been discussed already.

**Comment by coffeescriptbot** [comment]
_From @carlmathisen on July 27, 2016 9:44_

> @carlsmith: const x = 1 and const window.x = 1 are equivalent

I don't think that is the case, because `var x = 1` is scoped, and `x = 1` (without `var`, `let` or `const`) is assigned to `window`.

```
const x = 1;
alert(window.hello);
```

outputs `undefined`, while

```
x = 1;
alert(window.hello);
```

outputs `1`

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 27, 2016 12:13_

Thanks @carlmathisen. Yep, you're right. I never knew it worked like that. It just never came up. That makes things consistent at least. And the ability to make properties read-only makes `const`properties redundant, which wraps the whole thing up. Thanks again for putting me straight. I at least understand the decisions now.

ES6 gave JavaScript features we really needed, but also made a pretty messy language even messier :(

We ultimately need a new language, with JS bindings, but with its own semantics that refine JavaScript, removing cruft like the special arguments array type and 'temporal dead zones', and fixing coercion and exception handling etc. etc. Once WASM matures, it'll be much easier to do without the performance penalty of using an interpreter VM written in JavaScript instead of just compiling to JS.

**Comment by coffeescriptbot** [comment]
_From @DomVinyard on July 28, 2016 1:11_

> We ultimately need a new language, with JS bindings, but with its own semantics that refine JavaScript

I'm here because I need Coffeescript. And because it would be nice to get a community together again to lovingly support it going forward.

It's still my belief that there is no need for a var/let/const distinction in Coffeescript and that 'a variable is a variable' is enormously powerful.

**Comment by coffeescriptbot** [comment]
_From @JimPanic on July 28, 2016 4:09_

> ES6 gave JavaScript features we really needed, but also made a pretty messy language even messier :(

That's certainly true. I hope a few of those messy bits will get cleaned up over time in the next years, though. Given the pace ES is moving at right now, this is not too unlikely.

> It's still my belief that there is no need for a var/let/const distinction in Coffeescript and that 'a variable is a variable' is enormously powerful.

I think it's a good feature to have as long as you don't _have to_ use it. I do want to be able to use `const` if I deem necessary, though, and not have CS stand in the way.

**Comment by coffeescriptbot** [comment]
_From @carlmathisen on July 28, 2016 4:26_

But is it a core focus feature,  right now? It's absence is certainly not in the way of ES6 compatability, such as classes. 

IMO it's nice to have at best, not need to have.

**Comment by coffeescriptbot** [comment]
_From @JimPanic on July 28, 2016 4:32_

`const` is quite a low-hanging fruit, I think (but am not sure). Personally, for me it _is_ a need a to have, as opposed to classes which I don't use at all. So this is a matter of personal priorities and eventually the priorities we agree upon in the project. It does not seem to be a technical decision after all. :)

It is not listed as the proposed core focus features, but as one of points in the "four core may be too short"-list. Can you maybe head over to the #8 issue and comment there as well regarding priorities?

**Comment by coffeescriptbot** [comment]
_From @DomVinyard on July 28, 2016 8:03_

> Personally, for me it is a need a to have

I can fathom the sort of project where the overhead of const over var is a performance bottleneck remedy. I can't fathom why anybody would pick coffeescript for that project.

**Comment by coffeescriptbot** [comment]
_From @carlsmith on July 28, 2016 8:45_

> We ultimately need a new language, with JS bindings, but with its own semantics that refine JavaScript
> 
> > I'm here because I need Coffeescript. And because it would be nice to get a community together again to lovingly support it going forward.

I was making an observation about what the world _ultimately_ needs.  I don't see how your motivations for being here counter that.

> Personally, for me it is a need a to have, as opposed to classes which I don't use at all.

The ability to extend ES6 classes is required to use the APIs provided by modern JavaScript libraries. If you don't have that feature, you can't use a growing list of libraries. We don't need constants for ES6 interop. It's not about personal preferences. We need ES6 classes to avoid becoming obsolete, but constants are entirely optional.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on July 28, 2016 10:21_

`const` is a nice-to-have in my book as well, but @JimPanic is certainly not alone in wanting it. 

Let's talk priorities on the priorities thread, and avoid shooting down ideas just because they're not "absolute must-haves"

**Comment by coffeescriptbot** [comment]
_From @carlmathisen on July 28, 2016 9:29_

You're right @rattrayalex, sorry about the tangent.

**Comment by coffeescriptbot** [comment]
_From @GeoffreyBooth on Sep 6, 2016 12:03_

Carrying over a discussion from #30, is interoperability affected if we _don’t_ add support for `const`? Like will there be any library we can’t use if we can’t send a `const`-defined variable into it, for example? Or perhaps a build tool that optimizes based on `const` vs `var` or `let`?

One sort-of case I can think of is mentioned in [this comment](https://github.com/jashkenas/coffeescript/pull/4300#issuecomment-244855727) on the modules PR. Basically, anything you import is essentially a `const`. In this:

``` js
import foo from 'lib'
```

`foo` is read-only, meaning it’s essentially a `const`; if you try `foo = 2` later on, an error will be thrown. I wonder if it’s weird to have these implicit constants defined via `import`, while no other constants can be created in any other way. @greghuc, with regards to your ”mental model” of needing to keep track of which variables are read-only and which aren’t, unfortunately with module support you have no choice but to fight that battle (assuming you want to use modules).

**Comment by coffeescriptbot** [comment]
_From @zeekay on Sep 6, 2016 17:23_

> Any thoughts on const being used like so:
> x := 7  # const x = 7

I'm not a fan as `:=` for `const` declaration as it's used in Go (and other languages) for variable declaration. If we were going to break nonlocal assignment in CoffeeScript I'd prefer to use `:=` for new variables and `=` for assigning to an existing variable (as in Go). This makes assigning to a variable in an outer scope very clear and tends to actually prevent bugs whereas I've never found `const` to provide much value.

As `const` does seem to be fairly popular though, support makes sense. I'd argue that we should just we make `const` pass through like other ES6 constructs.

```
const x = 7
```

This is perfectly acceptable and matches JavaScript which is preferable to new syntax.
It's readable and fine, in my opinion.

Coming from Python, CoffeeScript's nonlocal assignment was kind of confusing, at least initially. I no longer find it very difficult to reason about, but I cannot think of a single person I've met who is a fan of it (and of course it's been written about for quite some time, never positively). Personally, I would not mind ditching it in favor of implicit `let` or explicit `:=`/`=`.

**Comment by coffeescriptbot** [comment]
_From @GeoffreyBooth on Sep 6, 2016 17:57_

zeekay I think `:=` was selected because `const` can only be assigned as part of its declaration, so it was appropriate. There is a strong convention in CoffeeScript to not use keywords for declaration (hence no `var`, no `function`, etc.; exceptions exist like `class`, but I don’t think people want to add any more). We can’t break backwards compatibility, so changing the behavior of `=` is not an option.

By nonlocal assignment, I assume you mean CoffeeScript’s habit of creating a `var` line at the top of whatever scope you’re in and declaring all variables there, then assigning them wherever you reference the variable? That won’t be possible with `const`: by definition, it needs to be assigned when it’s declared. I assume that support for `const` would put its declaration/assignment at the same place in the code that you typed it in CoffeeScript, rather than moving it up to the top of the scope.

**Comment by coffeescriptbot** [comment]
_From @zeekay on Sep 6, 2016 18:48_

By nonlocal assignment I was referring to how it's not possible to shadow an outer variable: 

```
x = 1
do ->
    x = 2 # x assigns to nonlocal x above
x # == 2
```

Which is the terminology used above in this discussion (as a nod to Python's `nonlocal` statement, I believe). The default of nonlocal assignment is very confusing to newcomers and does lead to bugs, in my experience.

I would argue that the lack of keywords for declaration was due to lack of necessity (when CoffeeScript was introduced there was only `var` to be concerned with). CoffeeScript has actually introduced a great number of keywords (`loop`, `do`, `is`, `isnt`, etc). I tend to find keywords easier to read and reason about compared to new syntactic structures and that's been one of the things I've come to appreciate about CoffeeScript.

**Comment by coffeescriptbot** [comment]
_From @zeekay on Sep 6, 2016 18:58_

As a quick follow-on: I'd be fine with changing declaration behavior. Compiling normal declarations directly to `let` would lead to an improvement in the readability of the compiled JavaScript and more closely match what modern JavaScript developers would write by hand.

The more interesting bit of the discussion in my mind is what to do (if anything) about CoffeeScript's (for lack of better terminology) nonlocal assignment. It might be nice to introduce `:=` as a way to force declaration, which would not break backwards compatibility:

```
x = 1
do ->
    x := 2 # declares a new x
x # == 1
```

Which would be a better use of `:=` than as shorthand for `const`.

**Comment by coffeescriptbot** [comment]
_From @GeoffreyBooth on Sep 6, 2016 23:52_

@zeekay may I invite you to open an issue suggesting changing nonlocal assignment? And the community can debate its merits there.

This thread covers both explicit `const` assignment (the proposed `:=` operator) and automatic usage of `let` when appropriate. These are really two separate feature requests, especially considering that the former can be implemented today without a flag (since it’s opt-in by using the new syntax) while the latter can’t (since it’s automatic). I created https://github.com/coffeescript6/discuss/issues/31 to cover the proposed new operator. Automatic `let`s are much farther in our future, when we’re onto the task of converting all of CoffeeScript’s output to be as ES-latest as possible, so I’m not going to bother creating an issue for it (yet). But feel free if you want to discuss it further.

**Comment by coffeescriptbot** [comment]
_From @mrmowgli on Sep 7, 2016 16:46_

I can see why the const becomes important, especially when dealing with code security. Let being the counterpart to avoid intentional abuse of the browser environment. I don't think anything would directly break, but I do think there is a good reason to include the functionality. Perhaps there are some intelligent rules to creating let assignments automatically. +1 for := assuming let and keeping the const keyword.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on Sep 7, 2016 21:33_

@GeoffreyBooth not sure if anyone answered your question:

>  is interoperability affected if we don’t add support for const? 

I don't believe so, no. `const` is a feature people like, but not an interoperability concern.

**Comment by coffeescriptbot** [comment]
_From @carlsmith on Sep 8, 2016 13:54_

I can't see any way it could undermine interoperability either. It's the reference that is constant, not the value, and a library wouldn't even see the name, unless you were sharing globals. Every interface is focussed on the values, never the way they're expressed.

**Comment by coffeescriptbot** [comment]
_From @rattrayalex on Sep 9, 2016 9:10_

My understanding of our consensus: 
1. We'll probably just keep compiling to `var` for the time being, as their is no practical benefit of using `let` or `const` in the output. 
2. `const` may be implemented with `:=` or something similar, but won't be inferred; discussion for that has moved to #31 .
3. At some point in the future, coffeescript may compile to `let`/`const` instead of `var`, but it's not a priority and not worth discussing now. 
4. The `decaffeinate` project can be used if you want ES6 output today.

As such, I am closing this issue for now. If I get sufficient 👎 reactions to this comment, we'll re-open.

**Comment by coffeescriptbot** [comment]
_From @GeoffreyBooth on Sep 9, 2016 9:22_

If I could amend that consensus a little, @zeekay described an alternate proposal on Gitter that he’s planning to present as a new issue soon, probably this weekend. Basically his point is that the great advantage of `let`/`const` is its block-scoping, i.e.:

``` js
let a = 1;
if (true) {
  let b = 2;
}
console.log(b); // undefined
```

This is a dramatic improvement over `var`, and a big reason why `let` and `const` have become popular features. We need to at least discuss if this is something CoffeeScript should have, separate from the feature of `const` that means “throw an error on reassignment.”

I think we could have both, via `:=` and perhaps `:==` operators (or whatever two operators people think are best):
- `a := 1` would mean, “declare and assign `a` with the value of 1, using `let` if `a` gets reassigned in this block and `const` if it never gets reassigned.”
- `a :== 1` would mean, “declare and assign `a` with the value of 1 using `const`. If it gets reassigned later, throw an error.”

We don’t necessarily need the second operator, if we don’t care to give people a way to force `const`. I’m not sure how popular the “throw an error on reassignment” feature is, but I suspect it’s quite popular.

Agreed that we should close this issue though, in favor of #31 and whatever new issue is created to discuss this “lexical assignment operator”.
